test/testrelative.c

```diff
@@ -39,21 +45,63 @@ static void loop(void)
     while (SDL_PollEvent(&event)) {
         SDLTest_CommonEvent(state, &event, &done);
         switch (event.type) {
+        case SDL_EVENT_KEY_DOWN:
+            if (event.key.key == SDLK_C) {
+                /* If warp emulation is active, showing the cursor should turn
+                 * relative mode off, and it should re-activate after a warp
+                 * when hidden again.
+                 */
+                if (SDL_CursorVisible()) {
+                    SDL_HideCursor();
+                } else {
+                    SDL_ShowCursor();
+                }
+            }
+            break;
         case SDL_EVENT_MOUSE_MOTION:
         {
-            mouseX += event.motion.xrel;
-            mouseY += event.motion.yrel;
+            rect.x += event.motion.xrel;
+            rect.y += event.motion.yrel;
+
+            /* Warp the mouse back to the center of the window if warp mode is active.
+             *
+             * NOTE: DO NOT DO THIS IN REAL APPS/GAMES!
+             *
+             *       This is an outdated method of handling relative pointer motion, and
+             *       may not work properly, if at all, on some platforms. It is here *only*
+             *       for testing the warp emulation code path internal to SDL.
+             *
+             *       Relative mouse mode should used instead!
+             */
+            if (warp) {
+                SDL_Window *window = SDL_GetMouseFocus();
+
+                if (window) {
+                    int w, h;
+                    float cx, cy;
+
+                    SDL_GetWindowSize(window, &w, &h);
+                    cx = (float)w / 2.f;
+                    cy = (float)h / 2.f;
+
+                    SDL_WarpMouseInWindow(window, cx, cy);
+                }
+            }
         } break;
         default:
             break;
         }
     }
+
+    SDL_Log("Done");
```

Comment:
Debug logging? Also, in warp mode there should be an initial warp to the center.

Reply:
Whoops, yes. That's gone.

It also now initially warps when the window gains input focus.